File: engineering/codebase/smartqr.frontend-services/src/domain/codes/content/types/EmailContent.ts
import { ContentType } from "../registry";

/** Defines the static mailto content — a recipient plus optional subject and body. */
export interface EmailContent {
  /** Gets or sets the content-type discriminator. */
  type: typeof ContentType.Email;

  /** Gets or sets the recipient email address. */
  to: string;

  /** Gets or sets the optional prefilled subject line. */
  subject?: string;

  /** Gets or sets the optional prefilled message body. */
  body?: string;
}

File: engineering/codebase/smartqr.backend-services/SmartQr.Domain/Codes/Content/Email/Models/EmailContent.cs
namespace SmartQr.Domain.Codes.Content.Email.Models;

/// <summary>Pre-filled email — encodes to a <c>mailto:</c> URI with optional form-encoded subject / body query params.</summary>
public sealed record EmailContent : CodeContent
{
    /// <summary>Recipient address.</summary>
    public required string To { get; init; }

    /// <summary>Optional subject line.</summary>
    public string? Subject { get; init; }

    /// <summary>Optional message body.</summary>
    public string? Body { get; init; }

    /// <inheritdoc />
    public override string Encode()
    {
        var parameters = new List<string>();

        var subject = ContentEncoding.Clean(Subject);
        if (subject.Length > 0)
            parameters.Add($"subject={ContentEncoding.FormUrlEncode(subject)}");

        var body = ContentEncoding.Clean(Body);
        if (body.Length > 0)
            parameters.Add($"body={ContentEncoding.FormUrlEncode(body)}");

        var query = string.Join("&", parameters);
        return $"mailto:{ContentEncoding.Clean(To)}{(query.Length > 0 ? $"?{query}" : string.Empty)}";
    }
}
